### PR TITLE
Update buildQueryPlan for new Query Plan format

### PR DESCRIPTION
**Pull request by JakeDawkins**
This PR builds off of #4410 

That PR defines the query plan types and the queryPlanExecutor, and this PR updates the query plan _builder_. 

Rather than refactoring the internals of every utility function in the query planner, I focused on two places: the fetch nodes and the selection sets. These two places are pretty much the leaves of the query plan and the only places that were actually changed by the type updates.

The important things to look at in this PR are the changes to the `buildQueryPlan` file and the serializers. I decided to update the serializers to work with the new query plan format, so we still have a pretty-printable interface for the query plan, and so snapshots across the codebase wouldn't have to all be manually reviewed 😅. The serializer updated added a newline after fragments, but I figure that's a small price to pay.

Note: Since I didn't update the entire query planner internals, and added some mappers at the end, I _do_ expect this query planner to be marginally **slower** than the existing one. It was either this, or undergo a massive refactor. I think this is an okay tradeoff, since a refactor of that complexity (which I was previously working on) could introduce some pretty nasty untested bugs or edge cases we hadn't previously seen.

todo:

- [x] query planner builds new format
- [x] planner tests are updated with new format
- [x] rest of gateway code complies with new format & tests pass
- [x] serialize new format to old human readable one in non-cucumber tests

### CODE REVIEW

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 1]
```diff
@@ -1,28 +1,28 @@
-import {
+  import {
```
Comment (abernix) [inline packages/apollo-gateway/src/QueryPlan.ts:1]: Unintentional indent?

```suggestion
import {
```

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 2]
```diff
@@ -38,20 +38,63 @@ export interface ParallelNode {
...
+
+export function getResponseName(node: FieldNode): string {
+  return node.alias ? node.alias : node.name;
+}
+
+export const trimSelectionNodes = (selections: readonly GraphQLJSSelectionNode[]): SelectionNode[] => {
+  let remapped: SelectionNode[] = [];
+
+  // intentionally leaving out fragment spread from the input selections, because I don't
+  // THINK that's possible to see in the final query plan -- prove me wrong
```
Comment (abernix) [inline packages/apollo-gateway/src/QueryPlan.ts:81]: If it's not clear right now, we should probably write a test scenario to verify:

1. To maintain that behavior.
2. To point to in validating this. (and so we can remove any doubt / this comment).

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 3]
```diff
@@ -38,20 +38,63 @@ export interface ParallelNode {
...
+  selections.map(selection => {
+    if(selection.kind === Kind.FIELD){
+      remapped.push({
+        kind: Kind.FIELD,
+        name: selection.name.value,
+        selections: selection.selectionSet ? trimSelectionNodes(selection.selectionSet?.selections) : undefined
```
Comment (abernix) [inline packages/apollo-gateway/src/QueryPlan.ts:87]: Could write this without the ternary to the same effect I think, but curious if you're being explicit?

```suggestion
        selections: selection.selectionSet && trimSelectionNodes(selection.selectionSet.selections)
```

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 4]
```diff
@@ -38,20 +38,63 @@ export interface ParallelNode {
...
+      });
+    }
+    if(selection.kind === Kind.INLINE_FRAGMENT){
+      remapped.push({
+        kind: Kind.INLINE_FRAGMENT,
+        typeCondition: selection.typeCondition?.name.value,
+        selections: trimSelectionNodes(selection.selectionSet?.selections)
```
Comment (abernix) [inline packages/apollo-gateway/src/QueryPlan.ts:94]: Is `selection.selectionSet` guaranteed in this case because it's a fragment?

File: packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts [hunk 5]
```diff
@@ -71,23 +84,23 @@ function printNode(
...
+        //           indentationNext,
+        //           depth,
+        //           refs,
+        //           printer,
+        //         ),
+        //       )
+        //       .join(`\n${indentationNext}`) +
+        //     config.spacingOuter +
+        //     indentation
+        //   : '') +
```
Comment (abernix) [inline packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts:103]: Curious about the disabling of this and preserving it is intentional? (If so, please add a comment!)

File: packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts [hunk 6]
```diff
@@ -62,7 +65,17 @@ function printNode(
             indentationNext
           : '') +
         printer(
-          node.selectionSet,
+          visit(parse(node.operation), {
+            OperationDefinition: op => {
+              if(isQueryPlanOperation(op)) {
```
Comment (abernix) [inline packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts:70]: I don't quite understand what a "query plan operation" (an unnamed query that has `representations` as a variable?) or why it's special cased here.  Can you help me understand that?

Comment (trevor-scheer) [reply packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts:70]: Can we rename this predicate to `isEntitiesOperation`?

File: packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts [hunk 7]
```diff
@@ -142,3 +155,12 @@ function printNodes(
 
   return result;
 }
+
+function isQueryPlanOperation(op: OperationDefinitionNode) {
```
Comment (trevor-scheer) [inline packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts:159]: It's possible for a user to write a query that passes the requirements in this predicate, accidentally falling into the `true` case. Is there a way we can tag an entities operation a bit more clearly, like an additional key on the object? If no, we could find the `_entities` field in the document with a visitor and that would be sufficient since federation "owns" that field and users shouldn't be writing queries to it.

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 8]
```diff
@@ -1,28 +1,28 @@
...
   schema: GraphQLSchema;
   operation: OperationDefinitionNode;
   fragments: FragmentMap;
 };
 
+export interface QueryPlan {
+  kind: 'QueryPlan';
+  node?: PlanNode;
+}
+
```
[the remark this answers is not in the document]
Comment (JakeDawkins) [reply packages/apollo-gateway/src/QueryPlan.ts:25]: Nah, this wasn't intentional. My hunch is that it was a part of that messed up rebase 🙃

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 9]
```diff
@@ -38,20 +37,96 @@ export interface ParallelNode {
...
+export function serializeQueryPlanNode (k: string , v: any) {
+  switch(k){
+    case "selectionSet":
+    case "internalFragments":
+    case "loc":
+    case "arguments":
+    case "directives":
+    case "source":
+      return undefined;
```
[the remark this answers is not in the document]
Comment (JakeDawkins) [reply packages/apollo-gateway/src/QueryPlan.ts]: Yep! Although this PR removes that code, since it was an intermediate transform while we didn't have working query planner with the new format.

File: packages/apollo-gateway/src/__tests__/buildQueryPlan.test.ts [hunk 10]
```diff
@@ -919,13 +918,15 @@ describe('buildQueryPlan', () => {
                   ...__QueryPlanFragment_1__
                 }
               }
+
```
[the remark this answers is not in the document]
Comment (JakeDawkins) [reply packages/apollo-gateway/src/__tests__/buildQueryPlan.test.ts:921]: I spent a while trying to track this down, and I'm still not 100% sure what's going on. The extra newline is only occurring on FragmentDefinitions that I can tell, and from what I can debug, it's coming from the `print` function in graphql-js. The output from `print` in `packages/apollo-gateway/src/snapshotSerializers/astSerializer.ts` results in a double newline before fragment defintion nodes. I could add a replace to strip the newline out, but I'm not sure of any other implications of doing that (it passes tests locally, but still). 

```js
print(remapInlineFragmentNodes(value))
  .trim()
  .replace(/\n\n/g, '\n') // < ----- I could add this
  .replace(/\n/g, '\n' + indentation)
```

File: packages/apollo-gateway/src/QueryPlan.ts (hunk 4 above)
Comment (JakeDawkins) [reply packages/apollo-gateway/src/QueryPlan.ts:94]: Yep! Good catch :)

File: packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts (hunk 6 above)
Comment (JakeDawkins) [reply packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts:70]: Renamed (in next commit) 👍 @abernix This is to preserve the printing we do in the query plan printer right now. For all the fetch nodes to entities, we don't print the `_entities` fetch explicitly or the `representations` variable declarations. 

For example, on a fetchNode for a query plan, without removing this definition, this is the difference in what would be printed:

```diff
# with the node removed
- {
-   ... on User {
-     risk
-   }
- }

# without the node removed
+ {
+ _entities(representations: $representations) {
+   ... on User {
+     risk
+   }
+ }
```

File: packages/apollo-gateway/src/__tests__/buildQueryPlan.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (abernix) [reply packages/apollo-gateway/src/__tests__/buildQueryPlan.test.ts:921]: I think we agreed that this might be caused by changes to the printer that we are comfortable with so I think we can resolve this.

File: packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts (hunk 6 above)
Comment (abernix) [reply packages/apollo-gateway/src/snapshotSerializers/queryPlanSerializer.ts:70]: Thank you for the clarification, @JakeDawkins.

File: packages/apollo-gateway/src/QueryPlan.ts (hunk 2 above)
Comment (abernix) [reply packages/apollo-gateway/src/QueryPlan.ts:81]: 🤔  @JakeDawkins Can you double click on this (plausible) thread that @trevor-scheer is referring to?

Comment (trevor-scheer) [reply packages/apollo-gateway/src/QueryPlan.ts:81]: Going to mark this as resolved, since this is only ever called on the `requires`, not the `operation` as I had hypothesized. A comment to articulate this would be great!

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 11]
```diff
@@ -38,20 +38,73 @@ export interface ParallelNode {
...
+}
+
+/**
+ * Converts a GraphQL-js SelectionNode to our newly defined SelectionNode
+ *
+ * This function is used to remove the unneeded pieces of a SelectionSet's
+ * `.selections`. It is only ever called on a query plan's `requires` field,
+ * so we can guarantee there won't be any FragmentSpreads passed in. That's why
+ * we can ignore the case where `selection.kind === Kind.FRAGMENT_SPREAD`
+ */
```
Comment (abernix) [inline packages/apollo-gateway/src/QueryPlan.ts:84]: Thanks for clarifying this so well! 👍

File: packages/apollo-gateway/src/QueryPlan.ts [hunk 12]
```diff
@@ -38,20 +38,73 @@ export interface ParallelNode {
...
+export const trimSelectionNodes = (
+  selections: readonly GraphQLJSSelectionNode[],
+): SelectionNode[] => {
+  let remapped: SelectionNode[] = [];
```
Comment (abernix) [inline packages/apollo-gateway/src/QueryPlan.ts:88]: Could this be a `const` (If not, maybe a comment to explain, otherwise it doesn't look to be immediately clear to me how the re-assignment is happening).

**Review by abernix — APPROVED** [review]
Couple notes within, but LGTM to merge into #4410 and pickup over there! Thanks, @JakeDawkins!

File: packages/apollo-gateway/src/QueryPlan.ts (hunk 12 above)
Comment (JakeDawkins) [reply packages/apollo-gateway/src/QueryPlan.ts]: I think I originally did this as a signal that `remapped` was being mutated, at least in my mind, but there's no _need_ for the `const`. I swapped it out :)
